Color format adjustments (#257)
design-tokens/community-group: technical-reports/format/types.md

## Color

Represents a color in the UI. The `$type` property MUST be set to the string `color`. The `$value` property MUST be an object string containing a string `colorSpace`, a numeric `coordinates` array, optional numeric `alpha`, string `colorProfile`, and string `fallback` to support legacy color spaces as needed.

| Key            |      Type      | Required | Description                                                                                                                              |
| :------------- | :------------: | :------: | :--------------------------------------------------------------------------------------------------------------------------------------- |

> | `colorSpace`   |    `string`    |    Y     | An string representing the color space.                                                                                                  |

Reviewer: Note: we could probably link to all the valid values from CSS Color Module 4 (or possibly 5, if we want to extend to that)

```suggestion
| `colorSpace`   |    `string`    |    Y     | A string representing the color space.                                                                                                  |
```

> | `colorProfile` |    `string`    |          | String representing url of color profile.                                                                                                |

Reviewer: I think we are not using colorProfile, and just aligning with CSS Color Module 4? `colorSpace` is a shorthand for space + profile + gamut, also borrowing from the same spec

```suggestion
```

Author: In [CSS Color Module 4](https://drafts.csswg.org/css-color-4/#terminology) color terminology section, here's the definition of color space:

> A color space is an organization of colors with respect to an underlying colorimetric model, such that there is a clear, objectively-measurable meaning for any color in that color space. This also means that the same color can be expressed in multiple color spaces, or transformed from one color space to another, while still looking the same.

[Color profile](https://drafts.csswg.org/css-color-5/#at-profile) is different, it's about using a particular ICC color profile for device specific implementations (print, for example). For example, you can use the CMYK color space and choose more than one color profile to implement those colors.

Reviewer: You’re 100% right, and yes I do see us needing this. As we talked about earlier today, perhaps we limit this initial PR to CSS Color Module 4 where:

- only `colorSpace` is used
- `colorSpace` is limited to the allowed values in CSS Color Module 4 (`srgb`, `srgb-linear`, `display-p3`, etc.)

Then, we open a separate PR to expand the format to include CSS Color Module 5’s concept of `@color-profile`. That way we can get feedback on that separately. Not being prescriptive, but if we copy CSS Color Module 5’s mental model, I could see it introducing a new `$type: 'colorProfile'` token type that maps 1:1 to `@color-profile` (perhaps; maybe we think of another pattern!)

> | `coordinates`  | `number array` |    Y     | An array of numeric values representing individual color coordinates.                                                                    |

Reviewer: I think we were leaning toward “channels” being the term. Also just in the interest of being really specific, “an array” doesn’t specify the length of the list. Calling it a “tuple of 3 values” means that the values must always be 3—no more, no less. Just like any color in CSS, I think we should enforce that the length is always 3, no matter what.

```suggestion
| `channels`  | `[number, number, number]` |    Y     | Tuple of 3 numbers representing the 3 components of the color space. The allowed ranges depend on the `colorSpace` specified. |
```

Author: Yes, we were leaning towards that! 

The terms `component` and `channel` were used in different capacities in the W3C CSS Color Level docs, sometimes interchangeably. I agree that we should standardize terminology and clarify what this means.

For example in the [CSS Color Level 4](https://www.w3.org/TR/css-color-4/#color-type) color type definition it reads:

> Colors in CSS are represented as a list of color **components**, also **sometimes called “channels**”, representing axises in the color space. Each channel has a minimum and maximum value, and can take any value between those two. Additionally, every color is accompanied by an **alpha component**, indicating how transparent it is, and thus how much of the backdrop one can see through the color.

The [CSS Color Level 5](https://www.w3.org/TR/css-color-5/#intro) spec builds upon Level 4, yet the terminology is still interchangeable.

In [ColorJS](url), the term "coordinates" could be shorthand for component coordinate or channel coordinate if I assume correctly (need to confirm).

Happy to move back to `channels` for now. 

Also, I suggest we adjust wording to include `none` along with numbers in the allowable channel value types, to account for [4.4“Missing” Color Components and the none Keyword](https://www.w3.org/TR/css-color-4/#missing) support.

Reviewer 2: > Also, I suggest we adjust wording to include `none` along with numbers in the allowable channel value types, to account for [4.4“Missing” Color Components and the none Keyword](https://www.w3.org/TR/css-color-4/#missing) support.

Oh smart! I like the idea of aligning with all that great thinking as much as possible

Author: Yes, there have been interest in this support in the legacy color thread!

Reviewer: Oh also I think as I said in another comment, maybe we like `components` better and that’s fine with me! I don’t have any strong opinions. We could decide on the exact words at the last minute without it changing the proposal too much.

Reviewer 3: > Calling it a “tuple of 3 values” means that the values must always be 3—no more, no less.

This should be an array of arbitrary length.
CMYK for example has 4 channels.
Other print profiles might have many more for spot colors and varnishes.

Defining these as arrays now means that implementers are less likely to make assumptions that these will always have a length of 3.

> | `alpha`        |    `number`    |          | An integer or floating-point value representing the numeric value.                                                                       |

Reviewer: We should specify this MUST be in the range of `(0, 1)` (inclusive). `0` = 0%, `1` = 100%.

Author: Can do. Where should this sentence occur?

Reviewer: Anywhere in the `alpha` description

> | `fallback`     |    `string`    |          | Unit of distance. Supported values: [HEX 6](https://www.w3.org/TR/css-color-4/#typedef-hex-color) including the preceding `#` character. |

Reviewer: ```suggestion
| `hex`     |    `string`    |          | Fallback sRGB-8 color for colors using higher gamuts or higher bit depths. MUST be in the format of a [HEX 6](https://www.w3.org/TR/css-color-4/#typedef-hex-color) string, including the preceding `#` character. |
```

For example, initially the color tokens MAY be defined as such:

<aside class="example">

> ```json

Reviewer: Suggestion: if using comments, `jsonc` (“JSON with **C**omments) won’t show visible errors in syntax highlighting

````suggestion
```jsonc
````

{
  "Hot pink": {
    "$type": "color",
    "$value": {
      "colorSpace": "display-p3",

>       "colorProfile": "http://example.org/display-p3.icc",

Reviewer: Here and below, the format should match the updated syntax of `colorSpace` + `channels` + `alpha` (optional) + `hex` (optional)

```suggestion
     "channels": [0.92, 0.2, 0.97],
      "hex": "#ff00ff" // HEX 6 supported only
```

Author: Thanks for this suggestion! I should clarify why the optional `colorProfile` is listed here. I was thinking of the cases for print where a specific color profile should be associated with the color space for conversion. Perhaps we do a separate callout for this and use CMYK as the color space example.
